src/lib.rs

```diff
@@ -265,3 +308,46 @@ macro_rules! block {
         }
     };
 }
+
+pub struct NbFuture<Ok, Err, Gen: FnMut() -> Result<Ok, Err>> {
+    gen: Gen,
+}
+
+impl<Ok, Err, Gen: FnMut() -> Result<Ok, Err>> From<Gen> for NbFuture<Ok, Err, Gen> {
+    fn from(gen: Gen) -> Self {
+        Self { gen }
+    }
+}
+
+impl<Ok, Err, Gen: FnMut() -> Result<Ok, Err>> Future for NbFuture<Ok, Err, Gen> {
+    type Output = core::result::Result<Ok, Err>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
```

Comment:
hey thanks for this! given that the waker isn't getting propagated here i wouldn't expect the scheduler to poll this again, and always hitting wake runs into performance problems on non-embedded platforms... does this behave as expected / how're you using this at the moment?

(cc. @dirbaio the futures-whisper)

Reply:
Yes, it'll hang with most executors out there due to not waking the waker. Docs on [Future::poll()](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll) say supporting wakers is mandatory.

It'll only work with very dumb executors that do `loop { fut.poll(cx) }` ignoring the waker, pinning the CPU at 100%.

Reply:
There's one way around it, which is immediately waking ourselves on poll: `cx.waker().wake_by_ref()`. It'll still spin the CPU at 100% but at least it won't hang. In theory this might cause other tasks to get starved of CPU, but in practice executors usually handle a "self-wake" like this by moving the task to the back of the queue, so other tasks still get to run. I don't think it's guaranteed though.

IMHO the way forward is HALs implementing the async traits, and deprecate `nb`. This allows HALs to wake the wakers from irqs, avoiding spinning the cpu at 100%. It also allows using DMA (you [can't use DMA soundly with `nb`](https://github.com/rust-embedded-community/embedded-storage/issues/11#issuecomment-789701336)).

There's another issue: the way `nb` is currently used, the task is polled for each byte. This makes it too easy to lose data if the task isn't polled fast enough, as the hardware buffers are usually tiny. If HALs impl the futures instead, they have powerful tools to avoid that (irqs, dma).